### ISSUE TITLE
keyToDegree fails at computing a -0.5 degree

### ISSUE THREAD

**Issue by lgvr123** [supercollider/supercollider#5614]
## Environment

* SuperCollider version: 3.12.1
* Operating system: Win10

## Steps to reproduce

```supercollider
(-1).keyToDegree(Scale.mixolydian, 12);
```

## Expected vs. actual behavior
The expected result is -0.5.
The actual result is -1

(the same applies one octave up)

```keyToDegree``` computes correctly all the intermediate degrees expect for that one.

E.g.
```(-12,-11..12).collect(_.keyToDegree(Scale.mixolydian, 12))```
produces 
```[ -7.0, -6.5, -6.0, -5.5, -5.0, -4.0, -3.5, -3.0, -2.5, -2.0, -1, -1, 0.0, 0.5, 1.0, 1.5, 2.0, 3.0, 3.5, 4.0, 4.5, 5.0, 6, 6, 7.0 ]```

A degree -1 is computed 2 times, one for the note (-2) (correct) and one for (-1) incorrect.
A degree 6 is computed 2 times, one for the note 10 (correct) and one for 11 incorrect.

**Comment by jamshark70** [comment]
But...

```
(
[
	0,  // root = 0 OK
	0.1,  // root-sharp = 1 OK
	0.5,  // = -3 hahahahaha
	0.9,  // supertonic-flat = 1 OK
	1  // supertonic = 2 OK
].degreeToKey(a)
)
-> [ 0.0, 1.0, -3.0, 1.0, 2.0 ]

// therefore

1.keyToDegree(a).debug("c# degree")
.degreeToKey(a).debug("not c# anymore")

c# degree: 0.5
not c# anymore: -3.0
```

So I would not agree that the expected result of `(-1).keyToDegree(Scale.mixolydian, 12)` should necessarily be -0.5. There's a convincing case to be made that *all* of the chromatic alterations returned from keyToDegree are currently incorrect.

```
b = Scale.mixolydian.semitones;
-0.5.degreeToKey(b)
-> -5.0  // hahahaha
```

IMO the fact that degreeToKey and keyToDegree are not inverse operations is... at best messy.

The tricky thing is that B-natural in C mixolydian could be either a flatted tonic (-0.1) or a sharp 7th (-0.9). The easiest solution would be to return either a sharp or flat alteration always. A musically more useful solution would be to have some heuristics to determine which is more appropriate (but that would be more difficult, and slower).

degree --> key --> degree wouldn't be guaranteed to return the same number, but it would be an enharmonic equivalent.

I don't think we should change the definition of chromatics to divide the degree equally. If we did that, then harmonic minor would have F = 3, F# = 3.5, G = 4, Ab = 5, A = 5.33333, Bb = 5.6666667, B = 6, and at that point, diatonic transposition becomes extremely inconvenient (but, diatonic transposition is one of the most important reasons to represent pitches in terms of degrees).

Hence -- IMO -- I think keyToDegree is generally faulty for non-scale tones.

**Comment by telephon** [comment]
There are two possible interpretations of fractional values in a scale which collide here (accidental and linear). When I wrote `keyToDegree` I realised that the accidental-interpretation is very hard to handle, so instead of throwing away the information, I used a float for the linear deviation. Or my memory is wrong and the accidental-based interpretation came later, I don't know.

This comes up every once in a while, so perhaps a solution would be to add a linear variant of `degreeToKey` which is then the inverse of `keyToDegree`.

Would we have a use for that though? 

James McCartney had argued (at that time) that a linear mapping makes not much sense. But there may be cases where you really want to interpolate between scales in such a way.

**Comment by jamshark70** [comment]
> James McCartney had argued (at that time) that a linear mapping makes not much sense.

I agree with JMc on this. I would only disagree with his choice of 0.1 as the increment -- 0.01 would have made, for instance, 53-ET accidentals somewhat more useful.

In fact, I used to use linear division of scale steps, but then I ran into too many awkward cases in my live coding system, so I switched to the SC accidental approach. That is, my experience has been that the linear approach seems like a good idea, until you actually try to use it musically.

So my opinion is that keyToDegree and degreeToKey should be inverse operations according to the accidental scheme (understanding that keyToDegree may return a different enharmonic equivalent). If the linear variant is needed, let's have those as separate methods.

**Comment by lgvr123** [comment]
See also [keyToDegree is broken for fractional values in the scale step below the octave](https://github.com/supercollider/supercollider/issues/5741)
